xdsl/dialects/csl.py

```diff
@@ -883,6 +892,50 @@ def verify_(self) -> None:
             raise VerifyException(f"{self.name} can only operate on mem1d_dsd type")
 
 
+FunctionSignatures = list[
+    tuple[AttrConstraint | Attribute | type[Attribute] | TypeVar, ...]
+]
+
+
+class _BuiltinDsdOpBase(IRDLOperation, ABC):
+    ops = var_operand_def()
+
+    @staticmethod
+    @abstractmethod
+    def get_signatures() -> FunctionSignatures:
+        raise NotImplementedError()
+
+    def verify_(self) -> None:
+        def typcheck(
+            op_typ: Attribute,
+            sig_typ: AttrConstraint | Attribute | type[Attribute] | TypeVar,
```

Comment:
I am not sure we need `AttrConstraint` here, since we don't really use them anywhere else in the dialect (not directly any way). If you do think we should keep it, maybe consider using `attr_constr_coercion` from `irdl`, to convert everything to an  `AttrConstraint` then using the `verify` method.

Reply:
It's the type signature copied over from `operand_def`